dart-impl/mpi/src/dart_synchronization.c

```diff
@@ -25,252 +27,365 @@
+                   lock->teamid);
+    DART_ASSERT_RETURNS(dart__base__mutex_unlock(&lock->mutex), DART_OK);
     return DART_ERR_INVAL;
   }
 
+  int32_t predecessor, result;
+  dart_gptr_t gptr_tail = lock->gptr_tail;
+  dart_gptr_t gptr_list = lock->gptr_list;
 
-	/* MPI-3 newly added feature: atomic operation*/
-	MPI_Fetch_and_op(&unitid.id, predecessor, MPI_INT32_T, tail, offset_tail, MPI_REPLACE, dart_win_local_alloc);
-	MPI_Win_flush(tail, dart_win_local_alloc);
-
-  /* If there was a previous tail (predecessor), update the previous tail's next pointer with unitid
-   * and wait for notification from its predecessor. */
-  if (*predecessor != -1) {
-    if (dart_segment_get_disp(&team_data->segdata, seg_id, DART_TEAM_UNIT_ID(*predecessor), &disp_list) != DART_OK) {
-      return DART_ERR_INVAL;
-    }
-    win = team_data->window;
-
-		/* Atomicity: Update its predecessor's next pointer */
-		MPI_Fetch_and_op(&unitid.id, result, MPI_INT32_T, *predecessor, disp_list, MPI_REPLACE, win);
+  uint64_t    tail_offset = gptr_tail.addr_or_offs.offset;
+  dart_unit_t tail_unit   = gptr_tail.unitid;
 
-		MPI_Win_flush(*predecessor, win);
+  dart_team_unit_t unitid;
+  dart_team_myid(lock->teamid, &unitid);
 
-		/* Waiting for notification from its predecessor*/
-		DART_LOG_DEBUG("LOCK - waiting for notification from %d in team %d",
-				*predecessor, (lock -> teamid));
+  /* Fetch the current unit's tail and make this unit the new tail */
+  DART_ASSERT_RETURNS(
+    MPI_Fetch_and_op(
+      &unitid.id,
+      &predecessor,
+      MPI_INT32_T,
+      tail_unit,
+      tail_offset,
+      MPI_REPLACE,
+      dart_win_local_alloc),
+    MPI_SUCCESS);
+  DART_ASSERT_RETURNS(
+      MPI_Win_flush(tail_unit, dart_win_local_alloc),
+      MPI_SUCCESS);
+
+  /* If there was a previous tail (predecessor), update the previous tail's
+   * next pointer with unitid and wait for notification from its predecessor.
+   */
+  if (predecessor != -1) {
+    MPI_Win    win;
+    MPI_Status status;
+    int16_t    seg_id = gptr_list.segid;
+    MPI_Aint   disp_list;
+
+    DART_ASSERT(DART_OK == dart_segment_get_disp(
+          &team_data->segdata,
+          seg_id,
+          DART_TEAM_UNIT_ID(predecessor),
+          &disp_list));
+    win = team_data->window;
 
-    MPI_Recv(NULL, 0, MPI_INT, *predecessor, 0, team_data->comm,
+    /* Atomicity: Update its predecessor's next pointer */
+    DART_ASSERT_RETURNS(
+      MPI_Fetch_and_op(
+        &unitid.id,
+        &result,
+        MPI_INT32_T,
+        predecessor,
+        disp_list,
+        MPI_REPLACE,
+        win),
+      MPI_SUCCESS);
+
+    DART_ASSERT_RETURNS(
+      MPI_Win_flush(predecessor, win),
+      DART_OK);
+
+    /* Waiting for notification from its predecessor*/
+    DART_LOG_DEBUG("dart_lock_acquire: waiting for notification from "
+                   "%d in team %d",
+                   predecessor, lock->teamid);
+
+    MPI_Recv(NULL, 0, MPI_INT, predecessor, 0, team_data->comm,
```

Comment:
This is the same that @rkowalewski reported in #339. However, I cannot reproduce this locally. What is your test case? One node? Multi-node? 

I added two test cases to specifically test the `dart_lock` implementation: https://github.com/dash-project/dash/blob/bug-330-dart-lock/dash/test/dart/DARTLockTest.cc

Can you please run this basic test to verify correctness of DART log?